### ISSUE TITLE
GetDelegateForFunctionPointer_CollectibleType_ReturnsExpected seems to fail indeterminantly

### ISSUE THREAD

**Issue by hughbe** [dotnet/corefx#31271]
```c#
        [Fact]
        public void GetDelegateForFunctionPointer_CollectibleType_ReturnsExpected()
        {
            MethodInfo targetMethod = typeof(GetDelegateForFunctionPointerTests).GetMethod(nameof(Method));
            Delegate d = targetMethod.CreateDelegate(typeof(NonGenericDelegate));
            IntPtr ptr = Marshal.GetFunctionPointerForDelegate(d);

            AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Assembly"), AssemblyBuilderAccess.RunAndCollect);
            ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule("Module");
            TypeBuilder typeBuilder = moduleBuilder.DefineType("Type", TypeAttributes.Class | TypeAttributes.Public | TypeAttributes.Sealed | TypeAttributes.AnsiClass | TypeAttributes.AutoClass, typeof(MulticastDelegate));
            ConstructorBuilder constructorBuilder = typeBuilder.DefineConstructor(MethodAttributes.RTSpecialName | MethodAttributes.HideBySig | MethodAttributes.Public, CallingConventions.Standard, new Type[] { typeof(object), typeof(IntPtr) });
            constructorBuilder.SetImplementationFlags(MethodImplAttributes.Runtime | MethodImplAttributes.Managed);

            MethodBuilder methodBuilder = typeBuilder.DefineMethod("Invoke", MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.NewSlot | MethodAttributes.Virtual, targetMethod.ReturnType, targetMethod.GetParameters().Select(p => p.ParameterType).ToArray());
            methodBuilder.SetImplementationFlags(MethodImplAttributes.Runtime | MethodImplAttributes.Managed);

            Type type = typeBuilder.CreateType();

            Delegate functionDelegate = Marshal.GetDelegateForFunctionPointer(ptr, type);
            VerifyDelegate(functionDelegate, targetMethod);
        }
```

https://mc.dot.net/#/user/hughbe/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/7e273bfe4ea12870523d1c934d58000fb38a3d62/workItem/System.Runtime.InteropServices.Tests/analysis/xunit/System.Runtime.InteropServices.Tests.GetDelegateForFunctionPointerTests~2FGetDelegateForFunctionPointer_CollectibleType_ReturnsExpected
```
System.NotSupportedException : Delegate marshaling for types within collectible assemblies is not supported.
Stack Trace :
   at System.Runtime.InteropServices.Marshal.GetDelegateForFunctionPointerInternal(IntPtr ptr, Type t)
   at System.Runtime.InteropServices.Marshal.GetDelegateForFunctionPointer(IntPtr ptr, Type t) in /root/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:line 1637
   at System.Runtime.InteropServices.Tests.GetDelegateForFunctionPointerTests.GetDelegateForFunctionPointer_CollectibleType_ReturnsExpected() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetDelegateForFunctionPointerTests.cs:line 48

```

Not sure why this only fails from type to time?

**Comment by danmosemsft** [comment]
@AaronRobinsonMSFT

**Comment by AaronRobinsonMSFT** [comment]
My guess would be related to something with the GC collecting the delegate, but that is a guess. Note that the documentation is pretty clear here that the user needs to ensure the delegate isn't collected. Also, the exception states:

>Delegate marshaling for types within collectible assemblies is not supported

Which means when the assembly is created with `AssemblyBuilderAccess.RunAndCollect` it is collectible. Based on the code I can see the call to `Marshal.GetFunctionPointerForDelegate()` stores a value in a hashmap and in `Marshal.GetDelegateForFunctionPointer()` that same pointer is retrieved, but if that pointer isn't found for some reason - perhaps a GC? - then the type's metadata is queried which will result in the exception above.

I have tried a myriad of things to prove my GC theory, but alas no luck. I think trying this scenario with a GC stress would be enough to confirm or deny that the GC is involved. Otherwise, we probably need a consistent repro to get at the root cause.

**Comment by AaronRobinsonMSFT** [comment]
Any thoughts on this @luqunl?

**Comment by luqunl** [comment]
I agree this issue is related to GC.  Try to test add GC.KeepAlive()  in https://github.com/dotnet/corefx/pull/31392

**Comment by AaronRobinsonMSFT** [comment]
@hughbe Want to try @luqunl's fix and revert #31331?
